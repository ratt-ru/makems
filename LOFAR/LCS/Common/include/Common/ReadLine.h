//#  ReadLine.h: read a line from stdin using readline or cin
//#
//#  Copyright (C) 2009
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id: ReadLine.h 12536 2009-01-28 14:49:05Z diepen $

#ifndef LOFAR_COMMON_READLINE_H
#define LOFAR_COMMON_READLINE_H

// \file

// The functions in here read a line from stdin using readline.
// If readline is not found during configure, it reverts to reading from cin.
//
// By default file name completion can be used and command recall and editing.
// If context-sensitive completion is needed, the user should initialize
// readline explicitly.

//# Includes
#include <Common/lofar_string.h>

namespace LOFAR
{
  // Print a prompt, read a line, and add it to the history.
  // False is returned if EOF is given.
  // The resulting line is put in the argument \a line.
  bool readLine (string& line, const string& prompt=string(),
                 bool addToHistory=true);

  // Same as readLine, but remove leading whitespace, skip empty lines,
  // and optionally skip comment lines.
  // The default comment character string is empty meaning that no test
  // for comment lines is done.
  bool readLineSkip (string& line, const string& prompt=string(),
                     const string& commentChars=string());

} // namespace LOFAR

#endif
