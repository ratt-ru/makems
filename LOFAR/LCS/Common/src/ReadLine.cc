//#  ReadLine.cc: read a line from stdin using readline or cin
//#
//#  Copyright (C) 2009
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id: ReadLine.cc 12536 2009-01-28 14:49:05Z diepen $

#include <lofar_config.h>
#include <Common/ReadLine.h>
#include <Common/StringUtil.h>
#include <Common/lofar_iostream.h>
#ifdef HAVE_READLINE
# include <readline/readline.h>
# include <readline/history.h>
#endif

namespace LOFAR
{
#ifdef HAVE_READLINE
  bool readLine (string& line, const string& prompt, bool addToHistory)
  {
    char* str = readline(prompt.c_str());
    if (!str) return false;
    line = string(str);
    free(str);
    if (addToHistory) add_history (line.c_str());
    return true;
  }
#else
  bool readLine (string& line, const string& prompt, bool)
  {
    if (!prompt.empty()) cerr << prompt;
    getline (cin, line);
    return cin;
  }
#endif

  bool readLineSkip (string& line, const string& prompt,
                     const string& commentChars)
  {
    bool fnd = false;
    while (!fnd  &&  readLine (line, prompt, false)) {
      // Skip leading and trailing whitespace.
      uint st  = lskipws (line, 0, line.size());
      uint end = rskipws (line, st, line.size());
      uint sz = end-st;
      if (sz > 0) {                            
        // non-empty line.
        if (commentChars.empty()  ||  commentChars.size() > sz) {
          // Do not test for comment
          fnd = true;
        } else {
          for (uint i=0; i<commentChars.size(); ++i) {
            if (commentChars[i] != line[st+i]) {
              // non-comment
              line = line.substr(st, sz);
              fnd = true;
              break;
            }
          }
        }
      }
    }
#ifdef HAVE_READLINE
    add_history (line.c_str());
#endif
    return fnd;
  }

} // namespace LOFAR
