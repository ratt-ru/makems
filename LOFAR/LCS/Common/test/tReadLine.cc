//#  tReadLine.cc: Test program for the readLine functions
//#
//#  Copyright (C) 2009
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id: tReadLine.cc 12536 2009-01-28 14:49:05Z diepen $

#include <lofar_config.h>
#include <Common/ReadLine.h>
#include <Common/lofar_iostream.h>
#include <Common/lofar_fstream.h>

using namespace LOFAR;

int main (int argc, char* argv[])
{
  // Note that GNU readline writes on stdout.
  // Therefore output is written into a file which is handled in tReadLine.run.
  ofstream ostr("tReadLine_tmp.stdout");
  string line;
  if (argc <= 1) {
    while (readLine(line)) {
      ostr << line << endl;
    }
  } else {
    while (readLineSkip(line, "> ", argv[1])) {
      ostr << line << endl;
    }
  }
  return 0;
}
