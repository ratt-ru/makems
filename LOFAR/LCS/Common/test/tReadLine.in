string1

    
  str  
  //  aa
//
  / aa
