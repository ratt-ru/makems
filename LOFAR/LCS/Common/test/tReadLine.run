#! /bin/sh

# Note that GNU readline writes on stdout.
# Therefore tReadLine writes its output a file which is handled here.

echo "without skip:"
tReadLine < tReadLine.in > /dev/null
cat tReadLine_tmp.stdout
echo
echo "without comment skip:"
tReadLine "" < tReadLine.in > /dev/null
cat tReadLine_tmp.stdout
echo
echo "with comment skip:"
tReadLine "//" < tReadLine.in > /dev/null
cat tReadLine_tmp.stdout
