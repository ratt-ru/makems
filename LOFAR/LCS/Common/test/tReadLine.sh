#!/bin/sh
./runctest.sh tReadLine > tReadLine.log 2>&1
