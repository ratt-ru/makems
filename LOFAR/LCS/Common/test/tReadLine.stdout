without skip:
string1

    
  str  
  //  aa
//
  / aa

without comment skip:
string1
  str  
  //  aa
//
  / aa

with comment skip:
string1
str
/ aa
